src/tags.rs

```diff
@@ -1,17 +1,28 @@
-use crate::convert::{span_id_to_string, trace_id_to_string};
+use crate::{convert::{span_id_to_string, trace_id_to_string}, models::context_tag_keys::TAG_KEY_LOOKUP};
 use crate::models::context_tag_keys::{self as tags, Tags};
-use opentelemetry::{
-    exporter::trace::SpanData,
-    trace::{SpanId, SpanKind},
-};
+use opentelemetry::{exporter::trace::SpanData, trace::{SpanId, SpanKind}};
 use opentelemetry_semantic_conventions as semcov;
 
 pub(crate) fn get_tags_for_span(span: &SpanData) -> Tags {
     let mut map = Tags::new();
 
+    // First, allow the user to explicitly express tags with attributes that start with `ai.`
+    // These attributes do not collide with any opentelemetry semantic conventions, so it is
+    // assumed that the user intends for them to be a part of the `tags` portion of the envelope.
+    let ai_tags_iter = span.attributes.iter().filter(|a| a.0.as_str().starts_with("ai."));
+    for ai_tag in ai_tags_iter {
+        if let Some(ctk) = TAG_KEY_LOOKUP.get(ai_tag.0.as_str()) {
+            map.insert(
+                ctk.clone(), 
+                ai_tag.1.to_string()
+            );
+        }
+    }
+
+    // Set the operation id and operation parent id.
     map.insert(
         tags::OPERATION_ID,
-        trace_id_to_string(span.span_context.trace_id()),
+        span_id_to_string(span.span_context.span_id()),
```

Comment:
@frigus02, I made this change with the express purpose of asking you what is up here.

Why would you want the operation id to be the trace id?  Shouldn't the trace id be global through the entire execution of the application?  It seems to me like each span is an operation.  Right?

Reply:
I'm pretty sure `ai.operation.id` should map to the TraceId. The [data model for telemetry correlation](vsce publish --baseImagesUrl $REPO/extension-syntax-only/) says:

> To associate telemetry with a logical operation, every telemetry item has a context field called `operation_Id`. This identifier is shared by every telemetry item in the distributed trace. So even if you lose telemetry from a single layer, you can still associate telemetry reported by other components.

This is only satisfied by the TraceId. Imagine a distributed application where A calls B calls C. If B fails to send telemetry for some reason, then we can still correlate spans from A and C if both send the same operation id.

It then goes on to say:

> A distributed logical operation typically consists of a set of smaller operations that are requests processed by one of the components. These operations are defined by request telemetry. Every request telemetry item has its own `id` that identifies it uniquely and globally. And all telemetry items (such as traces and exceptions) that are associated with the request should set the `operation_parentId` to the value of the request `id`.
>
> Every outgoing operation, such as an HTTP call to another component, is represented by dependency telemetry. Dependency telemetry also defines its own id that's globally unique. Request telemetry, initiated by this dependency call, uses this `id` as its `operation_parentId`.

The description of smaller operations with their own id nicely fits spans and their SpanId. We can then use the parent span id in the `operation_parentId` field to link them together.

You can also have a look at some of the official Azure Exporters for OpenTelemetry, e.g. the Python one. It uses the same mapping:

https://github.com/microsoft/opentelemetry-azure-monitor-python/blob/2fc7e100653fc33cb480ffd5ff068edbe311bcec/azure_monitor/src/azure_monitor/export/trace/__init__.py#L74

Reply:
Interesting.  This may have been my own misunderstanding of trace_id.  It sounds like, even though the tracer is global, a new trace_id is generated per span.  Right?

Nonetheless, I will revert.

Reply:
Fixed.

Reply:
No, a new trace id is only generated for a new trace (i.e. a new root span). All child spans will then share the same trace id.

The naming in Application Insights is just super confusing, I think:

> A distributed logical operation typically consists of a set of smaller operations

So basically:

- distributed logical operation 👉 trace
- set of smaller operations 👉 spans

And the ai.operation.id corresponds to the distributed logical operation.

I think the operation name may not be set entirely correct at the moment, because we set it to something request specific here. Maybe we should inhert this from the root span. But we don't have access to the root span. So I don't think we have a choice.